src/persist-client/src/internal/state.rs

```diff
@@ -1194,13 +1195,19 @@ where
         let all_empty_reqs = merge_reqs
             .iter()
             .all(|req| req.inputs.iter().all(|b| b.batch.is_empty()));
-        if record_compactions && claim_unclaimed_compactions && all_empty_reqs {
+        if record_compactions && all_empty_reqs {
```

Comment:
I realize it was already like this, but somewhat nervous about CaAs with no updates claiming compactions. Thoughts on exempting those?

Reply:
I guess the question is - if we have a shard with data that needs compacting but no non-empty write traffic, do we want it to eventually be compacted?

So far I've been trying to get as many shards to be able to compact as I can manage. But I can see how this might introduce compaction for shards that would never otherwise see compactions of any kind, for better or for worse...

Are there concrete scenarios you're particularly worried about here?

Reply:
Concretely, I know that I've always assumed that "empty CaA" -> "no compaction work" and I've almost certainly said that to persist users, even if we've never written it down in the rustdoc. (It was also the entire basis for the SchemalessWriteHandle idea.) This assumption is what led to stuff like the placeholder schemas in txn-wal (and probably the shard finalization, though that wasn't me). We've fixed the placeholder schema stuff, but I don't recall if those are the only times I've made that assumption without like auditing all CaA callers...

I also think it breaks the "invariant" (probably more like north star than a hard guarantee) we were discussing yesterday that two CaA calls should be roughly equivalent to batching them together. That's also broken by not making these claimed requests amortized proportional to the number of updates in them like regular compaction work is, but somehow the non-empty case feels less bad

> So far I've been trying to get as many shards to be able to compact as I can manage.
Ftr, me too.

Reply:
Going to exclude empty appends here so I can get this merged, but we should put this on the agenda at some point... I think fundamentally the current maintenance model isn't extending well to the new guarantees we're trying to eke out of it!